### ISSUE TITLE
idea: Generic color names to streamline theme authoring

### ISSUE THREAD

**Issue by shenlebantongying** [hlissner/emacs-doom-themes#663]
tl;dr: emacs need an abstraction layer for colouring, so that new packages can be automatically supported. 

# Reason

Emacs lack colour names based on functionalities. In the same time, almost all the Emacs themes will define their own functionality-based colour names like `fg` `bg` `fg-alt` `fg-lighter` `functions` `vc-change` `comments` `strings`.....

To make themes work for more packages, theme authors have to coordinate colours for each package. For example, `doom-vibrant-theme` has special treatments for `centaur-tabs`, `elscreen`.... But, not all themes have nice special treatments for other themes.

For a causal customization themes user, when they try a new package, their customization doesn't fit with the pacakge's default!

A lot of complaints about Emacs themes, including doom-themes and many others, are theme X doesn't support package Y. Then theme author proceed to add some ad-hoc codes, but this job will never be finished!

This method certainly doesn't scale, as there will be more and more packages, plus tons of already exist packages. This endless need for supporting new packages should be avoided.

# Plan

Why not make a generic `doom-theme.el` that can plug generic colour names into each package's customization. The actual colours will be loaded from each theme's unique file. Then the theme authors can just assign colours for each functionality. Colours for other packages will be automatically supported.

We might have to aggregate what generic colours names got defined in `doom-themes`, and unify them.

https://github.com/hlissner/emacs-doom-themes/blob/e716ddbb882a3a06744faa74decb2fea1569c921/themes/doom-vibrant-theme.el#L91-L112

# Example:

`Apropospriate-theme` defined about  ~50 colours to unify two variants (light & dark) and many packages.
https://github.com/waymondo/apropospriate-theme/blob/e84255cf485004b6b2cb37ce9e0be62fd6512f8b/apropospriate.el#L46

**Comment by shenlebantongying** [comment]
The spacemacs theme also defined about ~50 generic names
https://github.com/nashamri/spacemacs-theme/blob/e5ed346b9c31f0b43eb359614efd9aa439e1d18d/spacemacs-common.el#L104

**Comment by shenlebantongying** [comment]
Sry,I wasn't reading the doom theme source code yet.
